db/view/view_building_worker.cc

```diff
@@ -0,0 +1,516 @@
+    }
+
+#ifdef SEASTAR_DEBUG
+    // validate that all tasks are in the same batch
+    for (auto& tid: ids) {
+        SCYLLA_ASSERT(_state.tasks_map[id]->tasks.contains(tid));
+    }
+#endif
+
+    if (_state.tasks_map[id]->state == view_building_worker::batch_state::idle) {
+        vbw_logger.debug("Starting batch with tasks {}", _state.tasks_map[id]->tasks);
+        if (!is_shard_free(_state.tasks_map[id]->replica.shard)) {
+            throw std::runtime_error(fmt::format("Tried to start view building tasks ({}) on shard {} but the shard is busy", _state.tasks_map[id]->tasks, _state.tasks_map[id]->replica.shard, _state.tasks_map[id]->tasks));
+        }
+        co_await _state.tasks_map[id]->start();
+    }
+
+    service::release_guard(std::move(guard));
+    while (!_as.abort_requested()) {
+        auto read_apply_mutex_holder = co_await _group0_client.hold_read_apply_mutex(_as);
+
+        if (are_tasks_finished()) {
+            co_return get_results();
+        }
+
+        // Check if the batch is still alive
+        if (!_state.tasks_map.contains(id)) {
+            throw std::runtime_error(fmt::format("Batch with task {} is not found in tasks map anymore.", id));
+        }
+
+        read_apply_mutex_holder.return_all();
+        co_await _state.tasks_map[id]->batch_done_cv.wait();
+    }
+    throw std::runtime_error("View building worker was aborted");
+}
+
+// Validates if the task can be executed in a batch on the same shard.
+static bool validate_can_be_one_batch(const service::view_building::view_building_task& t1, const service::view_building::view_building_task& t2) {
+    return t1.type == t2.type && t1.base_id == t2.base_id && t1.replica == t2.replica && t1.last_token == t2.last_token;
+}
+
+future<> view_building_worker::local_state::update(view_building_worker& vbw) {
+    const auto& vb_state = vbw._vb_state_machine.building_state;
+
+    // Check if the base table to process was changed.
+    // If so, we clear the state, aborting tasks for previous base table and starting new ones for the new base table.
+    if (processing_base_table != vb_state.currently_processed_base_table) {
+        co_await clear_state();
+        processing_base_table = vb_state.currently_processed_base_table;
+        vbw_logger.info("Processing base table was changed to: {}", processing_base_table);
+
+        if (processing_base_table) {
+            // When we start to process new base table, we need to flush its currrent data, so we can build the view.
+            
+            co_await vbw.container().invoke_on_all([base_id = *processing_base_table] (view_building_worker& local_vbw) -> future<> {
+                auto base_cf = local_vbw._db.find_column_family(base_id).shared_from_this();
+                co_await when_all(base_cf->await_pending_writes(), base_cf->await_pending_streams());
+                co_await flush_base(base_cf, local_vbw._as);
+            });
```

Comment:
We should avoid updating the in-memory state until all fallible operations are performed within the iteration of the worker. Here, you update the `processing_base_table` before you do the flush. If the flush fails (unlikely, but still can happen) then the worker will not retry it in the next iteration.

This is a real issue but I suppose it shouldn't be very common - let's fix this in a follow-up.